node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1074,71 +1075,182 @@ async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) ->
+    let batch_id =
+        BlockHeadersBatchId::from_known(lowest_trusted_block_header, MAX_HEADERS_BATCH_SIZE);
+
+    loop {
+        let fetched_headers_data: FetchedData<BlockHeadersBatch> =
+            fetch_retry_forever::<BlockHeadersBatch>(ctx, batch_id).await?;
+        match fetched_headers_data {
+            FetchedData::FromStorage { item } => {
+                return BlockHeadersBatch::validate(
+                    &*item,
+                    lowest_trusted_block_header,
+                    ctx.config.verifiable_chunked_hash_activation(),
+                )
+                .map_err(FetchBlockHeadersBatchError::InvalidBatchFromStorage)
+            }
+            FetchedData::FromPeer { item, peer } => {
+                match BlockHeadersBatch::validate(
+                    &*item,
+                    lowest_trusted_block_header,
+                    ctx.config.verifiable_chunked_hash_activation(),
+                ) {
+                    Ok(new_lowest) => {
+                        info!(?batch_id, ?peer, "received valid batch of headers");
+                        ctx.effect_builder
+                            .put_block_headers_batch_to_storage(item.into_inner())
+                            .await;
+                        return Ok(new_lowest);
+                    }
+                    Err(err) => {
+                        error!(
+                            ?err,
+                            ?peer,
+                            ?batch_id,
+                            "block headers batch failed validation. Trying next peer..."
+                        );
+                        ctx.effect_builder.announce_disconnect_from_peer(peer).await;
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Fetches blocks, their transactions and tries from Genesis until `trusted_block`.
+async fn fetch_blocks_and_state_since_genesis(ctx: &ChainSyncContext<'_>) -> Result<(), Error> {
+    info!("syncing blocks and deploys and state since Genesis");
+
+    let queue = Arc::new(WorkQueue::default());
+
+    // Flag set by a worker when it encounters an error.
+    let abort = Arc::new(AtomicBool::new(false));
+
+    for block_height in 0..=ctx.config.max_parallel_block_fetches() as u64 {
+        let block_hash = match ctx
+            .effect_builder
+            .get_block_hash_by_height_from_storage(block_height)
+            .await
+        {
+            Some(hash) => hash,
+            None => return Err(Error::NoSuchBlockHeight(block_height)),
+        };
+        queue.push_job(block_hash);
+    }
+
+    let mut workers: FuturesUnordered<_> = (0..ctx.config.max_parallel_block_fetches())
+        .map(|worker_id| fetch_block_worker(worker_id, abort.clone(), queue.clone(), ctx))
+        .collect();
+
+    while let Some(result) = workers.next().await {
+        result?; // Return the error if a download failed.
+        ctx.metrics.chain_sync_blocks_synced.inc();
```

Comment:
This will record the number of workers that finished, not the number of blocks synced. IIRC workers should exit only when there are no more jobs coming in.

Reply:
Good catch. Thanks.